cupy/sorting/search.py

```diff
@@ -49,9 +71,66 @@ def argmin(a, axis=None, dtype=None, out=None, keepdims=False):
     return a.argmin(axis=axis, dtype=dtype, out=out, keepdims=keepdims)
 
 
-# TODO(okuta): Implement nanargmin
+def nanargmin(a, axis=None):
+    """Return the indices of the minimum values in the specified axis ignoring
+    NaNs. For all-NaN slice ``ValueError`` is raised.
+    Subclass cannot be passed yet, subok=True still unsupported
+    Args:
+        a (cupy.ndarray): Array to take nanargmin.
+        axis (int): Along which axis to find the minimum. ``a`` is flattened by
+            default.
+    Returns:
+        cupy.ndarray: The indices of the minimum of ``a``
+        along an axis ignoring NaN values.
+    .. seealso:: :func:`numpy.nanargmin`
+    """
+    a, mask = _replace_nan(a, cupy.inf)
+    res = argmin(a, axis=axis)
+
+    if mask is not None:
+        mask = cupy.all(mask, axis=axis)
+        if cupy.any(mask):
+            raise ValueError('All-NaN slice encountered')
+
+    return res
```

Comment:
The parameter `a` can be assumed `cupy.ndarray` type (not array-like object), so we can implement more simply.
```py
if a.dtype.kind != 'f':
    return argmin(a, axis=axis)

mask = cupy.isnan(a)

if cupy.any(cupy.all(mask, axis=axis)):
    raise ValueError...

return argmin(cupy.where(mask, cupy.inf, a), axis=axis)
```

Reply:
you mean we can skip the whole replace_nan function by using this method in nanarg functions ?

Reply:
> so you mean we can skip the whole replace_nan function by using this method in nanarg functions?

Yes.

Reply:
Okay, so I have looked at the test_search.py and I am not sure as I haven't written any test cases before would like some help on writing them. Although, replacing that ```_replace_nan()``` with what you suggested makes the test_search.py fail in pytest run as it is without any changes.

##### Output of Pytest on test_search.py

_____________________ TestWhereTwoArrays_param_3.test_where_two_arrays ______________________
/usr/lib/python3.7/site-packages/cupy/testing/parameterized.py:60: in wrap
    six.raise_from(e_new.with_traceback(e.__traceback__), None)
/usr/lib/python3.7/site-packages/cupy/testing/parameterized.py:43: in wrap
    return method(*args, **kwargs)
/usr/lib/python3.7/site-packages/cupy/testing/helper.py:841: in test_func
    impl(self, *args, **kw_copy)
/usr/lib/python3.7/site-packages/cupy/testing/helper.py:122: in test_func
    accept_error=accept_error)
/usr/lib/python3.7/site-packages/cupy/testing/helper.py:57: in _check_cupy_numpy_error
    self.fail('Only cupy raises error\n\n' + cupy_tb)
E   AssertionError: Parameterized test failed.
E   
E   Base test method: TestWhereTwoArrays.test_where_two_arrays
E   Test parameters:
E     cond_shape: (3, 4)
E     x_shape: (2, 3, 4)
E     y_shape: (4,)
E   
E   AssertionError: Only cupy raises error
E   
E   Traceback (most recent call last):
E     File "/usr/lib/python3.7/site-packages/cupy/testing/helper.py", line 28, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/harshal/Works/main-cupy/cupy/tests/cupy_tests/sorting_tests/test_search.py", line 175, in test_where_two_arrays
E       return xp.where(cond, x, y)
E     File "/usr/lib/python3.7/site-packages/cupy/sorting/search.py", line 136, in where
E       return _where_ufunc(condition.astype('?'), x, y)
E     File "cupy/core/_kernel.pyx", line 833, in cupy.core._kernel.ufunc.__call__
E     File "cupy/core/_kernel.pyx", line 704, in cupy.core._kernel._guess_routine
E   TypeError: Wrong type ((<class 'numpy.bool_'>, <class 'numpy.complex128'>, <class 'numpy.int8'>)) of arguments for cupy_where
----------------------------------- Captured stdout call ------------------------------------
{'cond_type': <class 'numpy.float16'>, 'x_type': <class 'numpy.complex128'>, 'y_type': <class 'numpy.int8'>}

Reply:
Could you push your test to this branch? I'll take a look.

Reply:
I have pushed the changes as requested earlier, even with your suggested changes test cases fail and even with my first implementation it has failed.

Reply:
I ran the test but couldn't reproduce the above error. Could you try to rebuild cupy (`git clean -ffxd` and `pip install .`) and re-run the test?

Reply:
Any updates?

Reply:
Test cases are working as it now without any changes, I suppose there was some build issue at my end. I am still figuring out how to write test cases for the functions, I will give it a try again today and give the update on that.

Reply:
I have pushed some cython files could you please review them specially the file `_routines_statistics.pyx`

Reply:
`my_argmax_float` looks to choose `NaN` item with high priority. It is an unexpected behavior.
https://github.com/cupy/cupy/blob/46036c3bab5425f95087021c25bc1e485e39711d/cupy/core/_routines_statistics.pyx#L159-L160

Reply:
Got it. Removed the `is_nan()` look up.

Reply:
`1` is expected in the following case, but current implementation returns `0`.
```
>>> cupy.argmax(cupy.array([float('nan'), 0]))
array(0, dtype=int64)
```

Reply:
How about writing tests first?

Reply:
The travis CI is failing on windows for some indentation problem which I cannot reproduce on my end as I am on linux. Could you check it on your side? Also I have added `cupy.nanargmin` and `cupy.nanargmax` in sorting.rst  to create docs page but I keep getting no module named error for reasons unknown to me. Is there something more needed to be done?

Reply:
> How about writing tests first?

Alright, will do that first

> `1` is expected in the following case, but current implementation returns `0`.
> 
> ```
> >>> cupy.argmax(cupy.array([float('nan'), 0]))
> array(0, dtype=int64)
> ```

Oh right, thanks for that. Apologies for my naivety I'm not very familiar with cython.

Reply:
In my understanding, cupy.argmax(cupy.array([float('nan'), 0])) should be `0` as it doesn't ignores nan and cupy.nanargmax should return `1` right?

Reply:
Sorry, it's my typo.
```
>>> cupy.nanargmax(cupy.array([float('nan'), 0]))
array(0, dtype=int64)
```
`cupy.nanargmax` also returns `0`.

Reply:
I have fixed the issue now. The cython function now returns -1 when All-NaNs are encountered as I am not not sure how to raise an error from cython. I have also added the test cases. Please review and let me know.